File: advent-of-code/2020/Day12.swift
var directions = [(Character, Int)]()

while let line = readLine() {
    directions.append((line.first!, Int(line.dropFirst(1))!))
}

var first = 0
var shipAngle = 0

for (command, distance) in directions {
    if command == "N" || command == "W" || command == "F" && shipAngle > 90 {
        first += distance
    } else if command == "S" || command == "E" || command == "F" && shipAngle <= 90 {
        first -= distance
    } else if command == "L" {
        shipAngle = (shipAngle - distance + 360) % 360
    } else if command == "R" {
        shipAngle = (shipAngle + distance) % 360
    }
}

first = abs(first)

var waypoint = (x: -10, y: 1)
var ship = (x: 0, y: 0)

for (command, distance) in directions {
    if command == "N" {
        waypoint.y += distance
    } else if command == "S" {
        waypoint.y -= distance
    } else if command == "E" {
        waypoint.x -= distance
    } else if command == "W" {
        waypoint.x += distance
    } else if command == "L" && distance == 90 || command == "R" && distance == 270 {
        (waypoint.x, waypoint.y) = (waypoint.y, -waypoint.x)
    } else if command == "L" && distance == 270 || command == "R" && distance == 90 {
        (waypoint.x, waypoint.y) = (-waypoint.y, waypoint.x)
    } else if command == "F" {
        ship.x += distance * waypoint.x
        ship.y += distance * waypoint.y
    } else {
        waypoint.x *= -1
        waypoint.y *= -1
    }
}

let second = abs(ship.x) + abs(ship.y)

print("\(first) \(second)")

File: advent-of-code/2020/Day12.kt
import kotlin.math.abs

fun main() {
    val directions = mapLines { it[0] to it.drop(1).toInt() }

    var first = 0
    var angle = 0

    for ((direction, distance) in directions) {
        when {
            direction in "NW" || direction == 'F' && angle in 180..270 -> first += distance
            direction in "SE" || direction == 'F' && angle in 0..90 -> first -= distance
            direction == 'L' -> angle = (angle - distance + 360) % 360
            direction == 'R' -> angle = (angle + distance) % 360
        }
    }

    first = abs(first)

    var shipX = 0
    var shipY = 0
    var waypointX = -10
    var waypointY = 1

    for ((command, distance) in directions) {
        when {
            command == 'N' -> waypointY += distance
            command == 'S' -> waypointY -= distance
            command == 'E' -> waypointX -= distance
            command == 'W' -> waypointX += distance

            (command to distance).let { it == 'L' to 90 || it == 'R' to 270 } ->
                waypointX = waypointY.also { waypointY = -waypointX }

            (command to distance).let { it == 'L' to 270 || it == 'R' to 90 } ->
                waypointX = -waypointY.also { waypointY = waypointX }

            command == 'F' -> {
                shipX += distance * waypointX
                shipY += distance * waypointY
            }
            else -> {
                waypointX *= -1
                waypointY *= -1
            }
        }
    }

    val second = abs(shipX) + abs(shipY)

    println("$first $second")
}
